example/apollo-server/bookmarks.js

```diff
@@ -0,0 +1,66 @@
+import { makeAugmentedSchema } from '../../src/index';
+import { ApolloServer } from 'apollo-server';
+import neo4j from 'neo4j-driver';
+
+const typeDefs = `
+type Person {
+  _id: Long!
+  born: Int
+  name: String!
+  acted_in: [Movie] @relation(name: "ACTED_IN", direction: "OUT")
+  ACTED_IN_rel: [ACTED_IN]
+  directed: [Movie] @relation(name: "DIRECTED", direction: "OUT")
+  produced: [Movie] @relation(name: "PRODUCED", direction: "OUT")
+  wrote: [Movie] @relation(name: "WROTE", direction: "OUT")
+  follows: [Person] @relation(name: "FOLLOWS", direction: "OUT")
+  reviewed: [Movie] @relation(name: "REVIEWED", direction: "OUT")
+  REVIEWED_rel: [REVIEWED]
+}
+
+type Movie {
+  _id: Long!
+  released: Int!
+  tagline: String
+  title: String!
+  persons_acted_in: [Person] @relation(name: "ACTED_IN", direction: "IN")
+  persons_directed: [Person] @relation(name: "DIRECTED", direction: "IN")
+  persons_produced: [Person] @relation(name: "PRODUCED", direction: "IN")
+  persons_wrote: [Person] @relation(name: "WROTE", direction: "IN")
+  persons_reviewed: [Person] @relation(name: "REVIEWED", direction: "IN")
+}
+
+type ACTED_IN @relation(name: "ACTED_IN") {
+ from: Person!
+ to: Movie!
+ roles: [String]!
+}
+
+type REVIEWED @relation(name: "REVIEWED") {
+ from: Person!
+ to: Movie!
+ rating: Int!
+ summary: String!
+}
+
+`;
+
+const schema = makeAugmentedSchema({ typeDefs });
+
+const driver = neo4j.driver(
+  'bolt://localhost:7687',
+  neo4j.auth.basic('neo4j', 'letmein')
+);
+
+const server = new ApolloServer({
+  schema,
+  context: ({ req }) => {
+    return {
+      driver,
+      neo4jBookmarks: req.headers['neo4jbookmark']
```

Comment:
@johnymontana is there some related documentation we can follow? 
Sorry for my lack of Neo4j experience, so far I thought bookmarks were only used by read replica to stay up to date with the core servers in the context of the raft protocol. 

Reading up on [Causal chaining and bookmarks](https://neo4j.com/docs/driver-manual/current/cypher-workflow/index.html#driver-causal-chaining), it seems like the driver's session has a `lastBookmark()` method available. 
But I do not see a `context.neo4jBookmarks = context.driver.session().lastBookmark()` initialization in your code. 
Do I need to manually set ` req.headers['neo4jbookmark'] = context.driver.session().lastBookmark()`  then?

It seems like the example provided by the doc shows how to ensure that person nodes have been created before attempting to create a friendship relationship. It's difficult to imagine a similar real-life use case (a server would need to have access to both node information in order for a client to attempt a friendship connection with another).
What would be a practical use-case here? I initially thought about cursors and connections, but I suppose bookmarks are unrelated.

Reply:
The use case here is to support situations where a service other than neo4j-graphql.js is writing to the Neo4j cluster and we want the queries generated by neo4j-graphql.js to maintain the causal chain and only execute as of the last transaction to the cluster from that external service, as indicated by passing the bookmark returned from the external service's write transaction.

So you're correct in pointing out that this is only partial support for bookmarks in neo4j-graphql.js. What is missing is the ability to return a bookmark from a transaction executed via neo4j-graphql.js

Reply:
Makes sense now, thanks!